From 40167df65bbc78a4b8f2e955f5d56853a970702d Mon Sep 17 00:00:00 2001
From: Beniamino Galvani <bgalvani@redhat.com>
Date: Sun, 19 Jul 2026 22:57:45 +0200
Subject: [PATCH] dns/resolved: send the server port number

Send the DNS server port number to systemd-resolved when configured in
the URI (e.g. dns+udp://172.16.1.1:42).

Fixes: 4422b1470496 ('core, libnm: support per-connection DNS URIs')

https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/work_items/1869
(cherry picked from commit 5e2dd1bbfc5d580fc8ff7f73d8f114af9e63bd9a)
---
 src/core/dns/nm-dns-systemd-resolved.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/core/dns/nm-dns-systemd-resolved.c b/src/core/dns/nm-dns-systemd-resolved.c
index 172bd0bc6b..dcadf8fda5 100644
--- a/src/core/dns/nm-dns-systemd-resolved.c
+++ b/src/core/dns/nm-dns-systemd-resolved.c
@@ -331,7 +331,7 @@ call_done(GObject *source, GAsyncResult *r, gpointer user_data)
                 _LOGD("systemd-resolved support for SetLinkDNSEx(): API not supported");
 
                 _LOGW("systemd-resolved does not support SetLinkDNSEx API (v246). "
-                      "Cannot set DoT server name (SNI)");
+                      "Cannot set server port or DoT server name (SNI)");
 
                 /* We need to reconfigure with the SetLinkDNS fallback.
                  *
@@ -414,7 +414,7 @@ update_add_ip_config(NMDnsSystemdResolved    *self,
             continue;
         }
 
-        if (dns_server.servername) {
+        if (dns_server.port != NM_DNS_PORT_UNDEFINED || dns_server.servername) {
             NM_SET_OUT(out_require_dns_ex, TRUE);
             if (priv->has_set_link_dns_ex == FALSE) {
                 /* The caller won't care about this result anymore. We can skip setting it. */
@@ -428,7 +428,7 @@ update_add_ip_config(NMDnsSystemdResolved    *self,
             g_variant_builder_add_value(
                 dns_ex,
                 nm_g_variant_new_ay((gconstpointer) &dns_server.addr, addr_size));
-            g_variant_builder_add(dns_ex, "q", 0);
+            g_variant_builder_add(dns_ex, "q", dns_server.port);
             g_variant_builder_add(dns_ex, "s", dns_server.servername ?: "");
             g_variant_builder_close(dns_ex);
         }
@@ -531,7 +531,7 @@ prepare_one_interface(NMDnsSystemdResolved *self, const InterfaceConfig *ic)
     if (!require_dns_ex) {
         /* No need to use the new API. SetLinkDNS() is sufficient. */
     } else if (!priv->has_set_link_dns_ex) {
-        /* API to set server name is not supported. Nothing we can do. */
+        /* API to set port and server name is not supported. Nothing we can do. */
         require_dns_ex = FALSE;
     } else {
         g_variant_builder_init(&dns_ex, G_VARIANT_TYPE("(ia(iayqs))"));
-- 
GitLab

